compiler/rustc_middle/src/mir/mod.rs

```diff
@@ -1076,6 +1076,19 @@ pub struct VarDebugInfo<'tcx> {
 // BasicBlock
 
 rustc_index::newtype_index! {
+    /// The unit of the MIR [control-flow graph][CFG].
+    ///
+    /// There is no branching (e.g., `if`s, function calls, etc.) within a basic block, which makes
```

Comment:
Is this correct that there are no function calls within a basic block?

Reply:
Functions in general can "branch" by unwinding, but I don't know whether unwinding and non-unwinding functions are separated in rustc or not.

Reply:
Are function calls considered a branch? I think yes, because they're an unconditional jump. So that would mean that they can't occur within a basic block (as a statement), only as a terminator. Is that right?

Reply:
As far as I know this is correct -- calls are `TerminatorKind::Call` <https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/terminator/enum.TerminatorKind.html>, never part of an `Rvalue` <https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/enum.Rvalue.html>.

(That said, intrinsics may become special https://github.com/rust-lang/compiler-team/issues/348, but I think that doesn't affect the "no branching" statement.)

Reply:
> Is this correct that there are no function calls within a basic block?

Yes, this is correct and crucial for the notion of a "basic block". There cannot be any control dependencies inside a basic block.

Reply:
Although IIUC functions can be inlined in opt passes, so the content of the function is within one basic block but it’s not actually calling anything.

Reply:
Inlining will typically add many basic blocks -- one for each of the inlined function. Not sure what you mean by "the content of the function is within one basic block".

Reply:
I meant that if the function has no branching, when it's inlined I would think it could theoretically go into the middle of a basic block. Is that true? Or does MIR just always add a fresh basic block for an inlined function without checking if it can go into an existing block?

Here's an example of what I mean:

```rust
#[inline]
fn add(x: i32, y: i32) -> i32 { x + y }

fn add_two(x: i32) -> i32 { add(x, 2) }
```

If the compiler **does not** inline `add()`, then I'm guessing the CFG looks something like this (partially based on looking at Playground output):

```rust
fn add(_1: i32, _2: i32) -> i32 {
    let mut _0: i32;

    BB0 {
        _0 = _1 + _2;
        return;
    }
}

fn add_two(_1: i32) -> i32 {
    let mut _0: i32;

    BB0 {
        _0 = call add(_1, 2);
    }

    // Need a new basic block because of the call in BB0.
    BB1 {
        return;
    }
}
```

And if the compiler **does** inline `add()`, I would think that this is theoretically possible:

```rust
fn add_two(_1: i32) -> i32 {
    let mut _0: i32;

    // Everything can be in one basic block because the function was inlined.
    BB0 {
        _0 = _1 + 2;
        return;
    }
}
```

Is that roughly correct or does MIR add a fresh basic block whenever there's an inlined function call? I'm pretty new to CFGs and MIR, so I'm still trying to figure them out :). Thanks for your help in understanding them!

Reply:
> I meant that if the function has no branching, when it's inlined I would think it could theoretically go into the middle of a basic block. Is that true? Or does MIR just always add a fresh basic block for an inlined function without checking if it can go into an existing block?

There is a separate optimization pass that glues together multiple basic blocks when possible (specifically whenever there is a Goto from one BB to a BB that has just this one predecessor, then the two can be merged). So the inliner doesn't have to worry about this.

But I don't see how that's relevant for this PR.

Reply:
I just wanted to make sure that I understood how function-related stuff works in MIR. I’ll ask on the `t-compiler` stream on Zulip in the future to keep things on-topic :)

Thank you for clearing up my understanding! I really appreciate it.

Reply:
I see. :) Yes such discussion is better put on Zulip, also because there are more eyes watching and benefiting there. :)